src/Microsoft.OData.Core/UriParser/CustomUriFunctions.cs

```diff
@@ -142,18 +142,62 @@ public static bool RemoveCustomUriFunction(string functionName)
+        /// Output for the list of signature objects for matched function names, with canonical name of the function;
+        /// null if no matches found.
+        /// </param>
+        /// <param name="enableCaseInsensitive">Whether to perform case-insensitive match for function name.</param>
         /// <returns>true if the function was found, or false otherwise.</returns>
-        internal static bool TryGetCustomFunction(string name, out FunctionSignatureWithReturnType[] signatures)
+        internal static bool TryGetCustomFunction(string functionCallToken, out IList<KeyValuePair<string, FunctionSignatureWithReturnType>> nameSignatures,
+            bool enableCaseInsensitive = false)
         {
-            Debug.Assert(name != null, "name != null");
+            Debug.Assert(functionCallToken != null, "name != null");
 
             lock (Locker)
             {
-                return CustomFunctions.TryGetValue(name, out signatures);
+                IList<KeyValuePair<string, FunctionSignatureWithReturnType>> bufferedKeyValuePairs
+                    = new List<KeyValuePair<string, FunctionSignatureWithReturnType>>();
+
+                // Do case-sensitive search first.
+                if (CustomFunctions.ContainsKey(functionCallToken))
+                {
+                    FunctionSignatureWithReturnType[] signatureGroup = null;
+                    CustomFunctions.TryGetValue(functionCallToken, out signatureGroup);
+
+                    Debug.Assert(signatureGroup != null, "signatureGroup != null");
+
+                    foreach (FunctionSignatureWithReturnType sig in signatureGroup)
+                    {
+                        // case-sensitive match: canonical name is just the function token name.
+                        bufferedKeyValuePairs.Add(new KeyValuePair<string, FunctionSignatureWithReturnType>(functionCallToken, sig));
+                    }
+                }
+
+                // Continue to search for match, if case-insensitive is enabled.
+                if (enableCaseInsensitive)
+                {
+                    foreach (KeyValuePair<string, FunctionSignatureWithReturnType[]> func in CustomFunctions)
+                    {
+                        if (func.Key.Equals(functionCallToken, StringComparison.OrdinalIgnoreCase)
+                            /* Skip the exact match above to avoid double-counting. */
+                            && !func.Key.Equals(functionCallToken, StringComparison.Ordinal))
+                        {
+                            foreach (FunctionSignatureWithReturnType sig in func.Value)
+                            {
+                                // case-insensitive match: canonical name is the one from the CustomFunctions dictionary.
+                                bufferedKeyValuePairs.Add(new KeyValuePair<string, FunctionSignatureWithReturnType>(func.Key, sig));
+                            }
+                        }
+                    }
+                }

```

Comment:
Current code loops through CustomFunctions multiple times (once to find if the key exists, once to get the value if it does, and once again for case-insensitive, doing two comparisons for each item in the list). Wouldn't it be simpler (and more efficient, especially for default case of case insensitive) to replace 164-195 with something like:

                foreach (KeyValuePair<string, FunctionSignatureWithReturnType[]> func in CustomFunctions)
                {
                    if (func.Key.Equals(functionCallToken, enableCaseInsensitive ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal))
                    {
                        foreach (FunctionSignatureWithReturnType sig in func.Value)
                        {
                            bufferedKeyValuePairs.Add(new KeyValuePair<string, FunctionSignatureWithReturnType>(func.Key, sig));
                        }
                    }
                }